junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/Root.java

```diff
@@ -82,7 +95,13 @@ void prune() {
 	}
 
 	private boolean isExcluded(TestDescriptor descriptor, Filter<TestDescriptor> postDiscoveryFilter) {
-		return descriptor.getChildren().isEmpty() && postDiscoveryFilter.apply(descriptor).excluded();
+		FilterResult filterResult = postDiscoveryFilter.apply(descriptor);
+		boolean excluded = descriptor.getChildren().isEmpty() && filterResult.excluded();
+		if (excluded) {
+			exclusionReasonToDescriptorMap.computeIfAbsent(filterResult.getReason().orElse("Unknown"),
+				list -> new LinkedList<>()).add(descriptor);
+		}
```

Comment:
Agree, Logic should not hide inside excluded function.
Since Filter result is computed inside `isExcluded` method, what I can do it I can compute `FilterResult`  inside the `applyPostDiscoveryFilters` and pass it to isExcluded in place of passing postDiscoveryFilter. like` isExcluded(TestDescriptor descriptor, FilterResult filterResult) `.